async-hooks: add trace events to AsyncWrap

AndreasMadsen opened the pull request:
**rebase of https://github.com/nodejs/node/pull/14347** ~~This is for debugging purposes, I'm getting a segmentation fault when running just `node`, but with `trace_events` enables it works. @matthewloring doesn't get the segmentation fault, so I'm running the CI on this PR to check it out.~~

This will allow trace event to record timing information for all
asynchronous operations that are observed by AsyncWrap.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
async_hooks, trace_events

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10185/

matthewloring commented:
The CI is segfaulting as well. I can try to reproduce it again.

matthewloring commented:
I suspect this is a breakage inside the trace event macros. It has been a few versions of V8 since they were updated so we will probably have to do something like https://github.com/nodejs/node/pull/12127 again. As soon as this change lands, we'll be able to test for such breakages inside node to ensure the macros are updated when new versions of V8 are picked up.

matthewloring commented:
Actually it looks like the macros haven't changed much. There were some changes to how the tracing controller is maintained with the new platform implementation. Does this patch help?

```diff
diff --git a/src/node.cc b/src/node.cc
index d6678dd967..c14d5229dd 100644
--- a/src/node.cc
+++ b/src/node.cc
@@ -268,7 +268,7 @@ static struct {
         trace_enabled ? tracing_agent_->GetTracingController() : nullptr);
     V8::InitializePlatform(platform_);
     tracing::TraceEventHelper::SetTracingController(
-        trace_enabled ? tracing_agent_->GetTracingController() : nullptr);
+        trace_enabled ? tracing_agent_->GetTracingController() : new v8::TracingController());
   }
 
   void Dispose() {
```

AndreasMadsen commented:
@matthewloring it appears too.

Running a new CI: https://ci.nodejs.org/job/node-test-pull-request/10190/

AndreasMadsen commented:
I've added a simple `process.binding('trace_events')` module that just supports basic `emit` and `categoryGroupEnabled`. The latter is to see if a category is enabled, which we should use to check if the `node` category is enabled. _The code is a simplification of that in https://github.com/jasongin/nodejs/tree/tracejs._

There appears to be some bug with categories. Running the code below with `node --trace-events-enabled --trace-event-categories v8 benchmark_server.js 10` shows that both `v8` and `custom` are enabled. But only `v8` should be enabled. In the `trace_events` output the `custom` events are shown too, the reported category is `v8`.

If `node --trace-events-enabled --trace-event-categories custom benchmark_server.js 10` is used, then both `v8` and `custom` are reported disabled and there is no `trace_events` output.

/cc @jasongin @matthewloring

```javascript
'use strict';
const http = require('http');
const fs = require('fs');
const trace_events = process.binding('trace_events');

const sleeptime = parseInt(process.argv[2], 10) || 0;
console.log(`sleeping ${sleeptime} ms at each request`);

console.log(trace_events.categoryGroupEnabled('v8'));
console.log(trace_events.categoryGroupEnabled('custom'));

function sleep (msec) {
  var start = Date.now();
  while (Date.now() - start < msec) {}
}

const server = http.createServer(function (req, res) {
  sleep(sleeptime);
  fs.createReadStream(__filename).pipe(res);
});


server.listen(8080, '127.0.0.1', function () {
  let left = 100;
  for (let i = 0; i < 100; i++) {
    trace_events.emit('S'.charCodeAt(0), 'custom', 'request-time', i + 1, 'triggerId', 10)
    http.get('http://127.0.0.1:8080', function (res) {
      res.resume();
      res.on('end', function () {
        trace_events.emit('F'.charCodeAt(0), 'custom', 'request-time', i + 1, 'triggerId', 20)
        left -= 1;
        if (left === 0) server.close();
      });
    });
  }
});
```

AndreasMadsen commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,172 @@
+    num_args = 1;
+    Local<Context> context = env->isolate()->GetCurrentContext();
+    printf("name [A]: %s\n", name); // REQUEST
+    arg_names[0] = Utf8Value(env->isolate(), args[4]).out();
```
@addaleax could you take a look? I don't understand why fetching the c-string of `args[4]` changes the value of `name`.

addaleax replied:
@AndreasMadsen The memory associated with `Utf8Value.out()` is tied to the lifetime of the `Utf8Value` itself – you can’t really use `GetName` like this, sorry.

(It sounds like what happens here is that when `name = GetName()` is called, the name string is converted to utf8, that memory is returned by `GetName()`, then freed because the `Utf8Value` goes out of scope; when you try to update `arg_names[0]` here, the same thing happens, and this just re-uses the memory allocated for `name`.)

AndreasMadsen replied:
Thanks, this does make sense to me. However, I'm not sure how to appropriately fix it. Can you help me?

addaleax replied:
The easiest way would be to just keep the `Utf8Value` around while you use the string (basically what you do [here](https://github.com/nodejs/node/pull/15538/files/35c21fb62a840efbecec0c8a498eb10d86570003#diff-5a0ddabd197fdff8886e9fb9d85ff024R34) – this is also what we do in most of the codebase, so grepping around should show enough examples).

The second-easiest thing would probably be to create `std::string`s from those `Utf8Values` and use them instead, e.g. `GetName()` could return a `std::string` and should start working :)

AndreasMadsen replied:
Okay. I was worried if the `if`-block scoping would cause some issues.

addaleax replied:
Oh – right, the block scope is an issue here. I think you can actually move the `Utf8Value` declaration to outside of the diff … if that doesn’t work, `std::string` is probably the way to go

addaleax commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,172 @@
+extern tracing::Agent* tracing_agent;
+
+// The tracing APIs require category groups to be pointers to long-lived
+// strings. Those strings are stored here.
```
Looking at the V8 source code, this doesn’t seem to be true? (https://github.com/nodejs/node/blob/master/deps/v8/src/libplatform/tracing/tracing-controller.cc#L168 & related places use string comparisons, not pointer comparisons)

AndreasMadsen replied:
Yeah, I'm confused too. This is where it is documented: https://github.com/nodejs/node/blob/master/src/tracing/trace_event_common.h#L123

addaleax replied:
Hm, okay … /cc @matthewloring

addaleax replied:
I really can’t make out a code path where this is actually needed … but it would be really good to have that verified by somebody who is familiar with the tracing code

AndreasMadsen replied:
Agreed. It is very confusing. I think `trace_events` runs in a separate thread, can that affect it somehow?

addaleax commented on src/async-wrap.cc:
```diff
@@ -606,7 +627,33 @@ AsyncWrap::AsyncWrap(Environment* env,
+
+  if (env->destroy_async_id_list()->empty())
+    uv_timer_start(env->destroy_async_ids_timer_handle(),
+                   DestroyAsyncIdsCallback, 0, 0);
```
Feel free to ignore, but common code style is to use braces if the body of an `if` statement longer than one line

addaleax commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,166 @@
+  if (args.Length() < 6 || (args[4]->IsUndefined() && args[5]->IsUndefined())) {
+    num_args = 0;
+  }
+  else if (args[4]->IsString() && args[5]->IsNumber()) {
```
(feel free to also ignore: common code style is to have `}` and `else (if)` on the same line

addaleax commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,166 @@
+  }
+  else if (args[4]->IsString() && args[5]->IsNumber()) {
+    num_args = 1;
+    Local<Context> context = env->isolate()->GetCurrentContext();
```
You can just use `env->context()`, and if you want to have `context` be a local variable I’d put it somewhere near the top (where `env` is declared as well)

AndreasMadsen commented:
I have fixed all issues, added more tests, fixed lint issues and squashed all commits.

**This is now ready for review :)**

CI: https://ci.nodejs.org/job/node-test-pull-request/10399/

AndreasMadsen commented:
/cc @nodejs/diagnostics

addaleax commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,163 @@
+using v8::Object;
+using v8::Value;
+
+extern tracing::Agent* tracing_agent;
```
Is this used somewhere?

addaleax commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,163 @@
+  return insertion.first->c_str();
+}
+
+static inline char GetPhase(Environment* env, const Local<Value>& arg) {
```
nit: You can pass `Local`s as `Local<Value>`, they are basically just pointers so copying them is effectively free

AndreasMadsen commented:
@matthewloring so it turns out `trace_events` isn't that much faster than just using `fs.write` and template strings to dump the data. Are we doing something wrong?

![benchmark-plot](https://user-images.githubusercontent.com/505333/31216592-780ade70-a9b4-11e7-999b-44bdc5fdde20.png)

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10409/

matthewloring commented:
@AndreasMadsen I'm sorry but I don't fully understand the graph. Are you comparing throughput with trace event active (and trace points inside async wrap) vs. throughput with fs.write calls inside the JS async wrap code?

AndreasMadsen commented:
Hmm, I'm not sure if throughput is the right word. It is the requests per second.

The 4 runs are:
* `baseline`: `node server.js`
* `naive async_hooks(all)`: `node -r log.js server.js`
* `trace_events async_hooks(all)`: `node --trace-events-enabled --trace-event-categories node.async_hooks server.js`
* `trace_events async_hooks(cpp)`: `node --trace-events-enabled --trace-event-categories node.async_hooks server.js` _but with the [JS async_hooks consumer](https://github.com/nodejs/node/pull/15538/files#diff-76290c41b43561605e14f33474b653dfR65) always disabled._

AndreasMadsen commented:
Fixed the benchmark. The results are more reasonable now.

![benchmark-plot-1](https://user-images.githubusercontent.com/505333/31254831-8e0a6e04-aa2a-11e7-885c-72cb557ff305.png)

---

The comment is collapsed, but could you also look at https://github.com/nodejs/node/pull/15538#discussion_r142421167

matthewloring commented:
I agree that not logging next ticks would skew the results. I would expect trace events to do substantially better than any logging done from JS but would have to do some profiling to have a more detailed picture.

AndreasMadsen commented:
The CI is almost green, I'm getting this error on `win2008r2-vs2017/vs2015` specifically?

```
Mismatched <anonymous> function calls. Expected exactly 1, actual 2.
    at Object.exports.mustCall (c:\workspace\node-test-binary-windows\COMPILED_BY\vs2015\RUNNER\win2008r2-vs2017\RUN_SUBSET\1\test\common\index.js:485:10)
    at Object.<anonymous> (c:\workspace\node-test-binary-windows\COMPILED_BY\vs2015\RUNNER\win2008r2-vs2017\RUN_SUBSET\1\test\parallel\test-async-wrap-uncaughtexception.js:14:33)
    at Module._compile (module.js:600:30)
    at Object.Module._extensions..js (module.js:611:10)
    at Module.load (module.js:521:32)
    at tryModuleLoad (module.js:484:12)
    at Function.Module._load (module.js:476:3)
    at Function.Module.runMain (module.js:641:10)
    at startup (bootstrap_node.js:187:16)
```

I'm not sure what could have caused that :/

AndreasMadsen commented:
running windows CI again: https://ci.nodejs.org/job/node-test-commit-windows-fanned/12263/

AndreasMadsen commented:
Ouch, looks like the error is legit. But I'm totally lost how this error could have any relation to this PR. It is also oddly specific to the Windows version and VS version, so I'm not sure there are any good reasons. I'm clueless.

AndreasMadsen commented:
/cc @nodejs/platform-windows - Please help :/

BridgeAR commented:
Ping @nodejs/platform-windows

AndreasMadsen commented:
According to #16210, a workaround on the affected machines might be in place.

CI: https://ci.nodejs.org/job/node-test-pull-request/10849/

AndreasMadsen commented:
Can I get some LGTMs on this?

AndreasMadsen commented:
Oh yes. We actually have some documentation in https://github.com/nodejs/node/blob/master/doc/api/tracing.md - I will expand on that.

AndreasMadsen commented:
I've added some basic documentation. We should properly redo the `trace_events` documentation. To begin with, it should be called `trace_events` not `tracing` :), we should also describe what each category does and what information it provides. But I think that is outside the scope of this PR.

mcollina commented:
@AndreasMadsen can you please open up a new issue for that?

Still LGTM on this one.

AndreasMadsen commented:
Opened issue https://github.com/nodejs/node/pull/15538#issuecomment-337859069

AndreasMadsen commented:
The windows issue has disappeared :D

jasnell commented on lib/internal/trace_events_async_hooks.js:
```diff
@@ -0,0 +1,67 @@
+// The behaviour is not only useful but the same as the events emitted using
+// the specific C++ macros.
+const BEFORE_EVENT = 'b'.charCodeAt(0);
+const END_EVENT = 'e'.charCodeAt(0);
```
extremely minor nit but why not just hard code the char code?

AndreasMadsen replied:
This is more readable. Otherwise, they are just magic numbers. And there is basically no extra cost.

jasnell commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,164 @@
+
+static inline char GetPhase(Environment* env, const Local<Value> arg) {
+  if (!arg->IsNumber()) {
+    env->ThrowTypeError("trace_events type must be a number.");
```
Can the type checks be moved up to JS and use `internal/errors` instead?

AndreasMadsen replied:
I'm not sure. We need to make sure that `GetPhase` doesn't have unexpected behavior for unexpected input, right? We are definitely doing this in a lot of other places, so if it is an issue we need to find a more general solution.

We can certainly create an extra abstraction layer as `internal/trace_events.js` and then repeat the type checking there.

AndreasMadsen replied:
Remember, this is not a public interface. If this error ever throws it is because there is a bug in node core. The API stability of the message shouldn't really be a concern.

jasnell replied:
Perhaps these should be `CHECK`s then?

jasnell commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,164 @@
+    args.GetReturnValue().Set(False(env->isolate()));
+  } else {
+    args.GetReturnValue().Set(True(env->isolate()));
+  }
```
If I am remembering correctly, I *believe* this could just be...

```c
args.GetReturnValue().Set((*category_group_enabled == 0));
```

That is, I *think* `Set()` accepts `bool` values and handles those appropriately. I could be wrong tho.

AndreasMadsen replied:
You are right https://v8docs.nodesource.com/node-8.0/da/da7/classv8_1_1_return_value.html#abbe0803e94a0eb36d04d9bd17ecb96a0

addaleax replied:
Yup – you don’t even need the double parentheses :)

AndreasMadsen commented:
/cc @nodejs/diagnostics 

This work was sponsored by nearForm so it would be nice with some non-nearForm sign-offs :D

jasnell commented:
> it would be nice with some non-nearForm LGTMs

heh.. and here I was getting ready to sign off on it ;-)

AndreasMadsen commented:
@jasnell please check again.

CI: https://ci.nodejs.org/job/node-test-pull-request/10900/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11420/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11423/

AndreasMadsen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11453/

addaleax commented on lib/internal/trace_events_async_hooks.js:
```diff
@@ -0,0 +1,67 @@
+// non-init events, use a map to manually map the asyncId to the type name.
+const typeMemory = new Map();
+
+// It is faster to emit trace_events directly from C++. Thus, thus happens in
```
typo: `Thus, this`

AndreasMadsen commented:
Was just about to merge. Then there was another merge conflict :D

CI: https://ci.nodejs.org/job/node-test-pull-request/11466/

AndreasMadsen commented:
landed in d217b2850efb9005819d55b697a37cbe5bd0003c

bnoordhuis commented on src/node_trace_events.cc:
```diff
@@ -0,0 +1,136 @@
+
+}  // namespace node
+
+NODE_MODULE_CONTEXT_AWARE_BUILTIN(trace_events, node::InitializeTraceEvents)
```
This should have been `NODE_BUILTIN_MODULE_CONTEXT_AWARE(...)` and probably breaks the shared/static library build.  I'll open a PR to rectify and remove `NODE_MODULE_CONTEXT_AWARE_BUILTIN()` while I'm at it.

AndreasMadsen replied:
Thanks 😂

bnoordhuis commented:
Back-porters, this should land along with https://github.com/nodejs/node/pull/17071.
